### ISSUE TITLE
Deadlock(?) with new modeline

### ISSUE THREAD

**Issue by dolda2000** [stumpwm/stumpwm#588]
I merged #577 a couple of days ago, and just had StumpWM hard freeze on me (quite annoyingly in a section that held an Xorg grab, so I had to slime in from another computer to debug the problem).

The main thread was waiting on a mutex held by one of the modeline refresher threads. Unfortunately, I never managed to get a backtrace of the refresher thread in question (because I accidentally triggered a condition that killed that thread), but since the main thread was waiting for its mutex, I would assume there was a deadlock.

This is the backtrace of the main thread where it was hung:
```
0: ("bogus stack frame")
1: (SB-THREAD::%%WAIT-FOR-MUTEX #<unavailable argument> #<unavailable argument> #<unavailable argument> #<unavailable argument> #<unavailable argument> #<unavailable argument>)
2: (SB-THREAD::%WAIT-FOR-MUTEX #<SB-THREAD:MUTEX owner: #<SB-THREAD:THREAD "Refresher of mode line #S(MODE-LINE
                          :SCREEN #S<screen #<XLIB:SCREEN :0.0 1600x1200x24 TRUE-COLOR>>
                          :HEAD #S(frame 1 NIL 0 0 1600 1200)
                          :WINDOW #<WINDOW :0 80001A>
                          :FORMAT *SCREEN-MODE-LINE-FORMAT*
                          :POSITION BOTTOM
                          :CONTENTS NIL
                          :CC #S(STUMPWM::CCONTEXT
                                 :SCREEN #S<screen #<XLIB:SCREEN :0.0 1600x1200x24 TRUE-COLOR>>
                                 :WIN #<WINDOW :0 80001A>
                                 :PX NIL
                                 :GC #<GCONTEXT :0 8388635>
                                 :DEFAULT-FG 8355711
                                 :DEFAULT-BRIGHT 12566463
                                 :DEFAULT-BG 3355443
                                 :FG NIL
                                 :BG NIL
                                 :BRIGHTP NIL
                                 :REVERSEP NIL
                                 :COLOR-STACK NIL
                                 :FONT #<FONT fixed :0 8388631>)
                          :HEIGHT 17
                          :FACTOR 1183/1200
                          :MODE STUMP
                          :THREAD NIL
                          :LOCK #<MUTEX (free) {1005AF3F13}>
                          :EVENT #<WAITQUEUE  {1005AF3F33}>)" RUNNING {1005B03663}> {1005AF3F13}> #<SB-THREAD:THREAD "main thread" RUNNING {1001E07003}> NIL NIL NIL NIL NIL NIL)
3: ((FLET #:WITHOUT-INTERRUPTS-BODY-359 :IN SB-THREAD::CALL-WITH-MUTEX))
4: (SB-THREAD::CALL-WITH-MUTEX #<CLOSURE (FLET SB-THREAD::WITH-MUTEX-THUNK :IN STUMPWM::REDRAW-MODE-LINE) {7FFFF65568FB}> #<SB-THREAD:MUTEX owner: #<SB-THREAD:THREAD "Refresher of mode line #S(MODE-LINE
                          :SCREEN #S<screen #<XLIB:SCREEN :0.0 1600x1200x24 TRUE-COLOR>>
                          :HEAD #S(frame 1 NIL 0 0 1600 1200)
                          :WINDOW #<WINDOW :0 80001A>
                          :FORMAT *SCREEN-MODE-LINE-FORMAT*
                          :POSITION BOTTOM
                          :CONTENTS NIL
                          :CC #S(STUMPWM::CCONTEXT
                                 :SCREEN #S<screen #<XLIB:SCREEN :0.0 1600x1200x24 TRUE-COLOR>>
                                 :WIN #<WINDOW :0 80001A>
                                 :PX NIL
                                 :GC #<GCONTEXT :0 8388635>
                                 :DEFAULT-FG 8355711
                                 :DEFAULT-BRIGHT 12566463
                                 :DEFAULT-BG 3355443
                                 :FG NIL
                                 :BG NIL
                                 :BRIGHTP NIL
                                 :REVERSEP NIL
                                 :COLOR-STACK NIL
                                 :FONT #<FONT fixed :0 8388631>)
                          :HEIGHT 17
                          :FACTOR 1183/1200
                          :MODE STUMP
                          :THREAD NIL
                          :LOCK #<MUTEX (free) {1005AF3F13}>
                          :EVENT #<WAITQUEUE  {1005AF3F33}>)" RUNNING {1005B03663}> {1005AF3F13}> NIL T NIL)
5: (STUMPWM::REDRAW-MODE-LINE #S(MODE-LINE :SCREEN #1=#S<screen #<XLIB:SCREEN #2=:0.0 1600x1200x24 TRUE-COLOR>> :HEAD #S(frame 1 NIL 0 0 1600 1200) :WINDOW #3=#<XLIB:WINDOW #2#:0 80001A> :FORMAT *SCREEN-MODE-LINE-FORMAT* :POSITION :BOTTOM :CONTENTS "[^Bweb^b]  0+AnandTech: Hardware News and Tech Reviews Since 19  1-ZipEntry (Java Platform SE 8 ) - Chromium ^R 2*extracting menthol - YouTube - Chromium^r  3-Haven and Hearth" :CC #S(STUMPWM::CCONTEXT :SCREEN #1# :WIN #3# :PX #<XLIB:PIXMAP #2#:0 80165B> :GC #<XLIB:GCONTEXT #2#:0 8388635> :DEFAULT-FG 8355711 :DEFAULT-BRIGHT 12566463 :DEFAULT-BG 3355443 :FG NIL :BG NIL :BRIGHTP NIL :REVERSEP NIL :COLOR-STACK NIL ...) :HEIGHT 17 :FACTOR 1183/1200 :MODE :STUMP :THREAD #4=#<SB-THREAD:THREAD "Refresher of mode line #S(MODE-LINE
                          :SCREEN #S<screen #<XLIB:SCREEN :0.0 1600x1200x24 TRUE-COLOR>>
                          :HEAD #S(frame 1 NIL 0 0 1600 1200)
                          :WINDOW #<WINDOW :0 80001A>
                          :FORMAT *SCREEN-MODE-LINE-FORMAT*
                          :POSITION BOTTOM
                          :CONTENTS NIL
                          :CC #S(STUMPWM::CCONTEXT
                                 :SCREEN #S<screen #<XLIB:SCREEN :0.0 1600x1200x24 TRUE-COLOR>>
                                 :WIN #<WINDOW :0 80001A>
                                 :PX NIL
                                 :GC #<GCONTEXT :0 8388635>
                                 :DEFAULT-FG 8355711
                                 :DEFAULT-BRIGHT 12566463
                                 :DEFAULT-BG 3355443
                                 :FG NIL
                                 :BG NIL
                                 :BRIGHTP NIL
                                 :REVERSEP NIL
                                 :COLOR-STACK NIL
                                 :FONT #<FONT fixed :0 8388631>)
                          :HEIGHT 17
                          :FACTOR 1183/1200
                          :MODE STUMP
                          :THREAD NIL
                          :LOCK #<MUTEX (free) {1005AF3F13}>
                          :EVENT #<WAITQUEUE  {1005AF3F33}>)" RUNNING {1005B03663}> :LOCK #<SB-THREAD:MUTEX owner: #4# {1005AF3F13}> ...) #<unused argument>)
6: (STUMPWM::UPDATE-ALL-MODE-LINES)
7: (STUMPWM::PROCESS-MAPPED-WINDOW #S<screen #<XLIB:SCREEN :0.0 1600x1200x24 TRUE-COLOR>> #<XLIB:WINDOW :0 7000004>)
8: ((LAMBDA (&REST #:G140 &KEY :PARENT :SEND-EVENT-P :WINDOW &ALLOW-OTHER-KEYS) :IN "/home/fredrik/sys/src/stumpwm/events.lisp") :PARENT #<XLIB:WINDOW :0 507> :SEND-EVENT-P NIL :WINDOW #<XLIB:WINDOW :0 7000004>)
9: (STUMPWM::HANDLE-EVENT :DISPLAY #<XLIB:DISPLAY :0 (The X.Org Foundation R11902000)> :EVENT-KEY :MAP-REQUEST :EVENT-CODE 20 :SEND-EVENT-P NIL :SEQUENCE 1284 :PARENT #<XLIB:WINDOW :0 507> :EVENT-WINDOW #<XLIB:WINDOW :0 507> :WINDOW #<XLIB:WINDOW :0 7000004>)
10: ((FLET SB-THREAD::WITH-RECURSIVE-LOCK-THUNK :IN XLIB:PROCESS-EVENT))
11: ((FLET #:WITHOUT-INTERRUPTS-BODY-387 :IN SB-THREAD::CALL-WITH-RECURSIVE-LOCK))
12: (SB-THREAD::CALL-WITH-RECURSIVE-LOCK #<CLOSURE (FLET SB-THREAD::WITH-RECURSIVE-LOCK-THUNK :IN XLIB:PROCESS-EVENT) {7FFFF6556F3B}> #<SB-THREAD:MUTEX "CLX Event Lock" owner: #<SB-THREAD:THREAD "main thread" RUNNING {1001E07003}>> T NIL)
13: ((FLET SB-IMPL::TIMEOUT-BODY :IN XLIB:PROCESS-EVENT))
14: (XLIB:PROCESS-EVENT #<XLIB:DISPLAY :0 (The X.Org Foundation R11902000)> :HANDLER #<FUNCTION STUMPWM::HANDLE-EVENT> :TIMEOUT 0 :PEEK-P NIL :DISCARD-P NIL :FORCE-OUTPUT-P T)
15: ((FLET #:WITHOUT-INTERRUPTS-BODY-387 :IN SB-THREAD::CALL-WITH-RECURSIVE-LOCK))
16: (SB-THREAD::CALL-WITH-RECURSIVE-LOCK #<CLOSURE (FLET SB-THREAD::WITH-RECURSIVE-LOCK-THUNK :IN STUMPWM::DISPATCH-ALL) {7FFFF65573AB}> #<SB-THREAD:MUTEX "CLX Event Lock" owner: #<SB-THREAD:THREAD "main thread" RUNNING {1001E07003}>> T NIL)
17: ((FLET STUMPWM::DISPATCH-ALL :IN "/home/fredrik/sys/src/stumpwm/stumpwm.lisp") #<XLIB:DISPLAY :0 (The X.Org Foundation R11902000)>)
18: ((:METHOD IO-LOOP (STUMPWM::SBCL-IO-LOOP)) #<STUMPWM::SBCL-IO-LOOP {1004EF0783}> :DESCRIPTION "StumpWM") [fast-method]
19: (STUMPWM::STUMPWM-INTERNAL-LOOP)
20: (STUMPWM::STUMPWM-INTERNAL ":0")
21: (STUMPWM ":0")
22: (SB-INT:SIMPLE-EVAL-IN-LEXENV (STUMPWM) #<NULL-LEXENV>)
23: (EVAL (STUMPWM))
24: (SB-IMPL::PROCESS-EVAL/LOAD-OPTIONS ((:EVAL . "(load #P\"/home/fredrik/sys/src/stumpwm/stumpwm.asd\")") (:EVAL . "(asdf:oos 'asdf:load-op :stumpwm)") (:EVAL . "(stumpwm:stumpwm)")))
25: (SB-IMPL::TOPLEVEL-INIT)
26: ((FLET #:WITHOUT-INTERRUPTS-BODY-77 :IN SB-EXT:SAVE-LISP-AND-DIE))
27: ((LABELS SB-IMPL::RESTART-LISP :IN SB-EXT:SAVE-LISP-AND-DIE))
```

I haven't read the new mode-line code, but it does seem strange to me that the main thread is calling `REDRAW-MODE-LINE` when there's a thread supposedly dedicated for the purpose, so I can't help but wonder if that's related.

**Comment by dolda2000** [comment]
Upon closer inspection, I see that `REDRAW-MODE-LINE` just signals the refresher thread to do its thing, so I guess that wasn't related in itself.

**Comment by PuercoPop** [comment]
This issue may be related

https://github.com/stumpwm/stumpwm-contrib/issues/126

**Comment by jorams** [comment]
This should be fixed by #599.

**Comment by dolda2000** [comment]
In case anyone is still working on the threaded modeline, this just happened to me again (haven't had a chance to restart with #599 merged yet), and I got much more useful info this time.

What happened was that an error happened on the modeline thread. I'm really not sure about what caused it, though. The error was this:
```
  There is no applicable method for the generic function
    #<STANDARD-GENERIC-FUNCTION STUMPWM:WINDOW-HEAD (2)>
  when called with arguments
    (#S(TILE-WINDOW "[redacted]" #x6E00009)).
```
But I have no idea why there would be no applicable method on `WINDOW-HEAD` for a `TILE-WINDOW`. The only reason I can think of for this happening is because I would have redeclared the class during runtime or something like that, but I'm pretty sure I haven't done that.

Either way, that caused the modeline refresh thread invoke the debugger, which then simply sat indefinitely waiting for interactive input. I don't really know what stream it would have been trying to read input from, though, since StumpWM was started by xdm and therefore shouldn't have had a terminal, and stdin connected to /dev/null, but either way that was what was happening.

For the record, this is the stacktrace that led up to the missing-method error:
```
 (SB-PCL::CALL-NO-APPLICABLE-METHOD
  #<STANDARD-GENERIC-FUNCTION STUMPWM:WINDOW-HEAD (2)>
  (#S(TILE-WINDOW "[redacted]" #x6E00009)))
 ((LAMBDA (STUMPWM::W) :IN STUMPWM::HEAD-WINDOWS)
  #S(TILE-WINDOW "[redacted]" #x6E00009))
 (REMOVE-IF-NOT
  #<CLOSURE (LAMBDA (STUMPWM::W) :IN STUMPWM::HEAD-WINDOWS) {1005057D1B}>
  (#S(TILE-WINDOW "..." #x1E00011)
   #S(TILE-WINDOW "..." #xA00137)
   #S(TILE-WINDOW "..." #x2400064)
   #S(TILE-WINDOW "..." #x1C00011)
   #S(TILE-WINDOW "..." #x2200064)
   #S(TILE-WINDOW "..." #x280003C)
   #S(TILE-WINDOW "..." #x260003C)
   #S(TILE-WINDOW "[redacted]" #x6E00009)))
 (STUMPWM::FMT-HEAD-WINDOW-LIST
   ....
```
Again, not sure if anyone is still working on the threaded mode-line, but it seems to me that a key take-away is that the refresher thread should have more gracious error handling.
